ipaserver/install/dsinstance.py

```diff
@@ -96,27 +102,22 @@ def schema_dirname(serverid):
     return config_dirname(serverid) + "/schema/"
 
 
-def remove_ds_instance(serverid, force=False):
-    """A wrapper around the 'remove-ds.pl' script used by
-    389ds to remove a single directory server instance. In case of error
-    additional call with the '-f' flag is performed (forced removal). If this
-    also fails, then an exception is raised.
+def remove_ds_instance(serverid):
+    """Call the lib389 api to remove the instance. Because of the
+    design of the api, there is no "force" command. Provided a marker
+    file exists, it will attempt the removal, and the marker is the *last*
+    file to be removed. IE just run this multiple times til it works (if
+    you even need multiple times ....)
     """
-    instance_name = ''.join([DS_INSTANCE_PREFIX, serverid])
-    args = [paths.REMOVE_DS_PL, '-i', instance_name]
-    if force:
-        args.append('-f')
-        logger.debug("Forcing instance removal")
 
-    try:
-        ipautil.run(args)
-    except ipautil.CalledProcessError:
-        if force:
-            logger.error("Instance removal failed.")
-            raise
-        logger.debug("'%s' failed. "
-                     "Attempting to force removal", paths.REMOVE_DS_PL)
-        remove_ds_instance(serverid, force=True)
+    logger.debug("Attempting to remove instance %s", serverid)
+    # Alloc the local instance by name (no creds needed!)
+    ds = DirSrv(verbose=True, external_log=logger)
+    ds.local_simple_allocate(serverid)
+
+    # Remove it
+    lib389_remove_ds(ds)
```

Comment:
``remove_ds_instance`` doesn't have a force option. How about you run it a second time when the first attempt has failed?

Reply:
I'll leave this decision on @Firstyear

Reply:
It doesn't need a force option, because every removal is a 'force'. More generally, this is designed as a state machine, where the *first* file touched is a marker of the install - the last file removed is the same marker.

As a result, if interrupted at ANY point during install or removal you can always attempt a removal, and we ONLY remove the marker iff all removal actions completed correctly. TBH I should add an fsync in between to be sure.

Anyway, that's why you don't need force, because "state machines and engineering".

Reply:
@Firstyear what happens when you already have a broken installation or the administrator accidentaly removes some part of configuration?

Reply:
Does not matter. Partial install, broken config, anything. So long as the marker file exists, we remove *everything*.

Reply:
Actually:
The run of `remove_ds_instance` is successful on FRawhide after my little `setattr` hack:
```
2018-03-12T10:51:19Z DEBUG Removing /etc/sysconfig/dirsrv-DOM-171-084-ABC-IDM-LAB-ENG-BRQ-REDHAT-COM
2018-03-12T10:51:19Z DEBUG Complete
2018-03-12T10:51:19Z DEBUG Instance removed correctly.
```

but running `ipa-server-install` afterwards results in 
```
  [error] FileExistsError: [Errno 17] File exists: '/etc/dirsrv/slapd-DOM-171-084-ABC-IDM-LAB-ENG-BRQ-REDHAT-COM/schema'
ipapython.admintool: ERROR    [Errno 17] File exists: '/etc/dirsrv/slapd-DOM-171-084-ABC-IDM-LAB-ENG-BRQ-REDHAT-COM/schema'
ipapython.admintool: ERROR    [Errno 17] File exists: '/etc/dirsrv/slapd-DOM-171-084-ABC-IDM-LAB-ENG-BRQ-REDHAT-COM/schema'
ipapython.admintool: ERROR    The ipa-server-install command failed. See /var/log/ipaserver-install.log for more information
ipapython.admintool: ERROR    The ipa-server-install command failed. See /var/log/ipaserver-install.log for more information
```